pkg/platform/kube/monitoring/test/function_test.go

```diff
@@ -0,0 +1,312 @@
+
+			// create the missing configmap
+			configMap, err = suite.KubeClientSet.CoreV1().ConfigMaps(suite.Namespace).Create(configMap)
+			suite.Require().NoError(err, "Failed to create configmap")
+
+			// wait for k8s to recover deployment from missing configmap error
+			expectedUnavailableReplicas := 0
+			err = common.RetryUntilSuccessful(3*time.Minute,
+				time.Second,
+				func() bool {
+					deploymentInstance := &appsv1.Deployment{}
+					suite.GetResourceAndUnmarshal("deployment",
+						kube.DeploymentNameFromFunctionName(functionName),
+						deploymentInstance)
+					suite.Logger.InfoWith("Waiting for deployment unavailable replicas to be zero",
+						"currentUnavailableReplicas", deploymentInstance.Status.UnavailableReplicas,
+						"expectedUnavailableReplicas", expectedUnavailableReplicas)
+
+					// wait until all replicas are available
+					return int(deploymentInstance.Status.UnavailableReplicas) == expectedUnavailableReplicas
+				})
+			suite.Require().NoError(err, "Failed to wait for replicas to become healthy")
+
+			// wait for monitoring
+			time.Sleep(waitOnMonitoringIntervals)
+
+			// ensure function monitoring did not recover the function from its recent deploy error
+			suite.GetFunctionAndExpectState(getFunctionOptions, functionconfig.FunctionStateError)
+
+			return true
+		}, func(deployResult *platform.CreateFunctionResult) bool {
+
+			// let interval occur at least once
+			time.Sleep(waitOnMonitoringIntervals)
+
+			// ensure function in ready state, deploy passes
+			suite.GetFunctionAndExpectState(getFunctionOptions, functionconfig.FunctionStateReady)
+			return true
+		})
+}
+
+func (suite *FunctionMonitoringTestSuite) TestRecoverErrorStateFunctionWhenResourcesAvailable() {
+	functionName := "function-recovery"
+	createFunctionOptions := suite.CompileCreateFunctionOptions(functionName)
+	getFunctionOptions := &platform.GetFunctionsOptions{
+		Name:      createFunctionOptions.FunctionConfig.Meta.Name,
+		Namespace: createFunctionOptions.FunctionConfig.Meta.Namespace,
+	}
+
+	suite.DeployFunction(createFunctionOptions, func(deployResults *platform.CreateFunctionResult) bool {
+
+		// get the function
+		function := suite.GetFunction(getFunctionOptions)
+
+		// ensure function is ready
+		suite.Require().Equal(functionconfig.FunctionStateReady, function.GetStatus().State)
+
+		// get function pod, first one is enough
+		pod := suite.GetFunctionPods(functionName)[0]
+
+		// get node name on which function pod is running
+		nodeName := pod.Spec.NodeName
+
+		// mark the node as unschedulable, we want to evict the pod from there
+		suite.Logger.InfoWith("Setting cluster node as unschedulable", "nodeName", nodeName)
+		_, err := suite.KubeClientSet.CoreV1().Nodes().Update(&v1.Node{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      nodeName,
+				Namespace: suite.Namespace,
+			},
+			Spec: v1.NodeSpec{
+				Unschedulable: true,
+			},
+		})
+		suite.Require().NoError(err, "Failed to set nodes unschedulable")
```

Comment:
that's a great way to test this but may effect test stability if controller/dashboard pods restart
It does seem to be the easiest way though without updating the podspec and risking a race with controller

Reply:
there are no controller/dashboard pods here. kube test suite uses controller and dashboard directly.